pennylane/devices/qubit/adjoint_jacobian.py

```diff
@@ -67,6 +54,8 @@ def adjoint_jacobian(tape: QuantumTape):
 
     Args:
         tape (.QuantumTape): circuit that the function takes the gradient of
+        state (TensorLike): the final state of the circuit; if not provided,
```

Comment:
I'm wondering if it makes more sense to put this on the tape itself. I think it's my preference, especially because it means we won't have to change the return type of `simulate` depending on the input (which I do not love, but that might be a me thing)

Reply:
Do you mean store the final state on the tape itself? I'm against doing that since non-simulator devices don't define a state and have no use for that.

Reply:
hmm. I know PennyLane is supposed to opt for functional, but I also think functional code is supposed to have a standard return/call shape and I fear that we're kinda straying from it. we're already passing the state around elsewhere in the code so I will probably abandon this, but it feels a little... reckless

Reply:
Alternatively, we can break `simulate` down into two helper functions: `get_final_state` and `measure_final_state`. Whenever we call `simulate` with `return_final_state=True`, we can instead just call `get_final_state` and `measure_final_state` directly. 

Actually, I prefer this over my current solution. 🤦

Reply:
ooo that's nice, I really like that! I'm sure it'll be nice to have a function that returns the pre-measurement state in other scenarios too

Reply:
I second this suggestion.  Having two different return options for `simulate` is less than ideal. 

In summary, we can do something like:

```
final_state, is_batched_state = get_final_state(tape)
results = measure_final_state(tape, final_state, is_batched_state)
jac = adjoint_jacobian(tape, final_state)
```

Reply:
(and `simulate` will just be the first two lines of that)